### PR TITLE
[SPARK-29778][SQL] pass writer options to saveAsTable in append mode

### PR DESCRIPTION
**Pull request by SpaceRangerWes**
### What changes were proposed in this pull request?

`saveAsTable` had an oversight where write options were not considered in the append save mode.

### Why are the changes needed?

Address the bug so that write options can be considered during appends.

### Does this PR introduce any user-facing change?

No

### How was this patch tested?

Unit test added that looks in the logic plan of `AppendData` for the existing write options.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
cc @brkyvz

**Comment by brkyvz** [comment]
ok to test

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2DataFrameSuite.scala [hunk 1]
```diff
@@ -125,4 +128,34 @@ class DataSourceV2DataFrameSuite
...
+  testQuietly("SPARK-29778: saveAsTable: append mode takes write options") {
+
+    var plan: LogicalPlan = null
+    val listener = new QueryExecutionListener {
+      override def onSuccess(funcName: String, qe: QueryExecution, durationNs: Long): Unit = {
+        plan = qe.analyzed
+      }
+      override def onFailure(funcName: String, qe: QueryExecution, error: Throwable): Unit = {}
+    }
+    spark.listenerManager.register(listener)
```
Comment (brkyvz) [inline sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2DataFrameSuite.scala]: Can we unregister the listener at the end with a `try - finally` please?

**Review by brkyvz — APPROVED** [review]
LGTM with one minor nit! Thanks @SpaceRangerWes

**Comment by brkyvz** [comment]
add to whitelist

**Comment by brkyvz** [comment]
retest this please

**Comment by gatorsmile** [comment]
ok to test

**Review by cloud-fan — APPROVED** [review]

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2DataFrameSuite.scala [hunk 2]
```diff
@@ -125,4 +128,39 @@ class DataSourceV2DataFrameSuite
...
+
+      val df = Seq((1L, "a"), (2L, "b"), (3L, "c")).toDF("id", "data")
+      df.write.option("other", "20").mode("append").saveAsTable(t1)
+
+      sparkContext.listenerBus.waitUntilEmpty()
+      plan match {
+        case p: AppendData =>
+          assert(p.writeOptions == Map("other" -> "20"))
+        case other =>
+          fail(s"Expected to parse ${classOf[AppendData].getName} from query," +
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2DataFrameSuite.scala]: nit. `query,"` -> `query, "`.

Comment (SpaceRangerWes) [reply sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2DataFrameSuite.scala]: @dongjoon-hyun I won't be able to get to this until Thursday evening. I'm willing to address it, but I didn't know how soon this needed to be merged.

cc: @brkyvz

Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2DataFrameSuite.scala]: I see, @SpaceRangerWes .

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM (except one minor stuff). Thank you for your first contribution, @SpaceRangerWes .

To committers:
I'm okay with addressing it during merging or in a follow-up PR.

**Comment by brkyvz** [comment]
LGTM as well. I think we can address in a follow up. There are a couple minor things I want to add to the same files.

**Comment by dongjoon-hyun** [comment]
Merged to master. Thank you, @SpaceRangerWes . You are added to the Apache Spark contributor group. Also, thank you, @HyukjinKwon , @brkyvz , @cloud-fan , @gatorsmile !

**Pull request closed without merging**